debian/rlvm.6
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH RLVM 6 "Sept 2014" "rlvm"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
rlvm \- RealLive interpreter clone
.SH SYNOPSIS
.B rlvm
[\fIOPTION\fR]... <path to directory containing a Gameexe.ini file>
.SH DESCRIPTION
.PP
rlvm is a Free Software reimplementation of the VisualArt's KK's
RealLive interpreter. It is meant to provide Linux and Apple users
with a compatible, portable interpreter to play VisualArts games,
specifically those released by Key.
.TP 
\fB\-\-font\fR
Specifies the path to a TrueType font to use instead of trying to autodetect the font
.SH SUPPORTED GAMES
.sp 1
Currently, rlvm can run the following games:
.IP \[bu] 2
Kanon
.IP \[bu]
Kanon (With NDT's English patch)
.IP \[bu]
AIR
.IP \[bu]
CLANNAD / CLANNAD Full Voice Edition
.IP \[bu]
Planetarian
.IP \[bu]
Planetarian (US Edition)
.IP \[bu]
Tomoyo After
.IP \[bu]
Little Busters
.IP \[bu]
Kud Wafter
.TP 0
In addition, there is partial support for:
.IP \[bu] 2
Princess Bride
.IP \[bu]
Princess Brave
.IP \[bu]
ALMA ~zutto soba ni~
.IP \[bu]
SNOW
.IP \[bu]
Some of the game demos from the Kinetic Novel website.
.SH FILES
.I ~/.rlvm/
.RS
All rlvm save files are kept in this directory.
.SH BUGS?
While rlvm is generally stable, emulation of the RealLive virtual machine is
incomplete.
.SH NOT A BUG
English patches that were not compiled with rlBabel will not line break
correctly. You can check if your patch is compiled with rlBabel and the text
encoding is set to "Western" by hitting [F1] at any time.
.SH AUTHOR
rlvm was written by Elliot Glaysher, with the help of Haeleth. Special thanks
to Jagarl for licensing all his RealLive file reading code under liberal
licenses, and for doing the first reverse engineering of the RealLive
system. Windows port and bug fixes made possible by Rasqual Twilight.
.PP
This manual page was written by Elliot Glaysher <glaysher@umich.edu>.


NEWS.TXT
(For full history, please see the internal git changelog)
-------------------------------------------------------------------------
Version 0.14:                                                [2014-10-04]

- Support for the US edition of planetarian.
- Support for Kud Wafter.
- Further attempts to smooth animations and text output.
- Internal code cleanups and beautification; ported to C++11.

-------------------------------------------------------------------------

Version 0.13.1:                                              [2013-12-15]

- Fix crash on the Clannad loading screen on certain OpenGL drivers.
- Fix order of font loading to prefer working fonts on Linux.

-------------------------------------------------------------------------

Version 0.13:                                                [2013-10-13]

- rlvm will now warn on startup when it detects Japanese save data, but
  English patched game files, and offer to reset the save data. (This
  problem has been the number one support request for years.)
- Much better support for Little Busters. Most graphical glitches during
  the Little Busters Refrain have been fixed.
- TCC tone curve effects have been reverse-engineered and implemented
  (thanks to lurkmoar)
- Sepia scenes (and other graphical filters) should look much better.
- Simple shake commands implemented (fancy per-layer shaking still
  unimplemented).
- Make animations smooth: data should be uploaded to the graphics card
  before an animation loop starts, not while the animation loop is
  running.
- Fixes finding system fonts on Linux

-------------------------------------------------------------------------

Version 0.12:                                                [2011-05-14]

- User interface polish:
  - Display fatal error messages in a native dialog to user instead of
    just printing to the terminal.
  - Improvements to directory selector; complains if a directory isn't a
    game on Mac, shows a GTK+ selector on Linux.
  - Commissioned icon and better desktop integration on Linux.
  - Japanese localizations (and localization support for other languages).
  - Add nice messages instead of crashing on some error conditions
    (trying to play a game we don't know the encoding for, trying to
    play an AVG32 or Siglus game, etc).
- Fix graphics corruption in in-game dialogs when a dialog is brought
  up, and then fullscreen mode activated.
- Smooth the output of text in rlBabel using games.
- Don't play voice samples while fast forwarding.

-------------------------------------------------------------------------

Version 0.11:                                                [2011-03-06]

- Fix graphical defects in Clannad
- Implement voice replay buttons in scrollback
- Implement the "Return to Previous Selection" feature
- Fix volume controls. Each channel has both a user controlled and a
  programmer controlled volume knob; these are different and should be
  multiplied together.
- Several efficiency fixes. rlvm no longer spends 10% of its runtime
  searching the disk for files and only uploads dirty rectangles to the
  graphics card instead of sending the whole surface on every change.
- Fix the location of the name window in most games.
- Multiple fixes to automode and configuring automode.

-------------------------------------------------------------------------

Version 0.10:                                                [2011-01-30]

- Support for Tomoyo After Memorial Edition.
- Voice support for Planetarian.
- Fix multiple errors in the handling of glosses.

-------------------------------------------------------------------------

Version 0.9:                                                 [2010-10-16]

- Preliminary support for Little Busters! No support for Ecstasy yet.
  - There are significant graphics glitches, but the entire game is
    completable to the end of the Refrain.
- Fixes saving the current graphics state in most games. Previously
  created save games can't be fixed, but new saved games should save
  their graphics state correctly.
- Fix regression in the original edition of CLANNAD's main menu.
- Hide unavailable options in select() calls.

-------------------------------------------------------------------------

Version 0.8:                                                 [2010-01-31]

- HIK animations have been reverse engineered and implemented:
  - Opening pan of the city in Planetarian now displays.
  - Many things that use HIK animations in the ALMA opening work.
- Weather objects have been reverse engineered and implemented and are
  used in pretty much every game I can find. They're used for snow,
  sakura petals, rain, balls of light, etc.
- Implement better line breaking for Japanese text.
- Fix color issues (blue displayed as red and vice versa) on some
  displays.
- Fixed multiple compatibility issues with the free demos on the Kinetic
  Novel website. Some demos are almost playable now; fully supporting
  them is an ongoing project.

-------------------------------------------------------------------------

Version 0.7:                                                 [2009-11-07]

- AIR Standard Edition can be completed.
- Princess Bride can be completed and Princess Brave maybe works.
- At least one route in ALMA can be completed.
- Fixes intermittent crashes after loading saved game.
- Prevent text boxes from being positioned offscreen.
- Implement shadows on text.
- Support KOE voice archives.
- Support loose ogg vorbis file voice patches (such as the one for
  KANON).
- Chinese and Korean rlBabel support compiled in by default. (I can't
  find any patches so this is untested).
- Large number of small graphical fixes, including faces in text boxes,
  visual appearance of #SELBTNs (such as in Fuuko's route), and various
  shading effects.

-------------------------------------------------------------------------

Version 0.6.3:                                               [2009-08-15]


src/platforms/sdl/rlvm.cc
// -*- Mode: C++; tab-width:2; indent-tabs-mode: nil; c-basic-offset: 2 -*-
// vi:tw=80:et:ts=2:sts=2
//
// -----------------------------------------------------------------------
//
// This file is part of RLVM, a RealLive virtual machine clone.
//
// -----------------------------------------------------------------------
//
// Copyright (C) 2006, 2007 Elliot Glaysher
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
//
// -----------------------------------------------------------------------

// We include this here because SDL is retarded and works by #define
// main(inat argc, char* agrv[]). Loosers.
#include <SDL.h>

#include <boost/program_options.hpp>
#include <iostream>
#include <string>

#include "platforms/sdl/sdl_rlvm_instance.h"
#include "systems/base/system.h"
#include "utilities/file.h"

namespace po = boost::program_options;
namespace fs = boost::filesystem;

using std::cerr;
using std::cout;
using std::endl;
using std::string;

// -----------------------------------------------------------------------

void printVersionInformation() {
  cout
      << "rlvm (" << GetRlvmVersionString() << ")" << endl
      << "Copyright (C) 2006-2014 Elliot Glaysher, et all." << endl << endl
      << "Contains code that is: " << endl
      << "  Copyright (C) 2006-2007 Peter \"Haeleth\" Jolly" << endl
      << "  Copyright (C) 2004-2006 Kazunori \"jagarl\" Ueno" << endl << endl
      << "This program is free software: you can redistribute it and/or modify"
      << endl
      << "it under the terms of the GNU General Public License as published by"
      << endl
      << "the Free Software Foundation, either version 3 of the License, or"
      << endl << "(at your option) any later version." << endl << endl
      << "This program is distributed in the hope that it will be useful,"
      << endl
      << "but WITHOUT ANY WARRANTY; without even the implied warranty of"
      << endl << "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the"
      << endl << "GNU General Public License for more details." << endl << endl
      << "You should have received a copy of the GNU General Public License"
      << endl
      << "along with this program.  If not, see <http://www.gnu.org/licenses/>."
      << endl << endl;
}

void printUsage(const string& name, po::options_description& opts) {
  cout << "Usage: " << name << " [options] <game root>" << endl;
  cout << opts << endl;
}

int main(int argc, char* argv[]) {
  SDLRLVMInstance instance(&argc, &argv);

  // -----------------------------------------------------------------------
  // Parse command line options

  // Declare the supported options.
  po::options_description opts("Options");
  opts.add_options()("help", "Produce help message")(
      "help-debug", "Print help message for people working on rlvm")(
      "version", "Display version and license information")(
      "font", po::value<string>(), "Specifies TrueType font to use.");

  po::options_description debugOpts("Debugging Options");
  debugOpts.add_options()(
      "start-seen", po::value<int>(), "Force start at SEEN#")(
      "dump-seen", po::value<int>(), "Dumps rlvm's internal parsing of SEEN#")(
      "load-save", po::value<int>(), "Load a saved game on start")(
      "memory", "Forces debug mode (Sets #MEMORY=1 in the Gameexe.ini file)")(
      "undefined-opcodes", "Display a message on undefined opcodes")(
      "count-undefined",
      "On exit, present a summary table about how many times each undefined "
      "opcode was called")("trace", "Prints opcodes as they are run)");

  // Declare the final option to be game-root
  po::options_description hidden("Hidden");
  hidden.add_options()(
      "game-root", po::value<string>(), "Location of game root");

  po::positional_options_description p;
  p.add("game-root", -1);

  // Use these on the command line
  po::options_description commandLineOpts;
  commandLineOpts.add(opts).add(hidden).add(debugOpts);

  po::variables_map vm;
  try {
    po::store(po::basic_command_line_parser<char>(argc, argv)
                  .options(commandLineOpts)
                  .positional(p)
                  .run(),
              vm);
    po::notify(vm);
  }
  catch (boost::program_options::multiple_occurrences& e) {
    cerr << "Couldn't parse command line: multiple_occurances." << endl
         << " (Hint: this can happen when your shell doesn't escape properly,"
         << endl
         << "  e.g. \"/path/to/Clannad Full Voice/\" without the quotes.)"
         << endl;
    return -1;
  }
  catch (boost::program_options::error& e) {
    cerr << "Couldn't parse command line: " << e.what() << endl;
    return -1;
  }

  // -----------------------------------------------------------------------

  po::options_description allOpts("Allowed options");
  allOpts.add(opts).add(debugOpts);

  // -----------------------------------------------------------------------
  // Process command line options
  fs::path gamerootPath;

  if (vm.count("help")) {
    printUsage(argv[0], opts);
    return 0;
  }

  if (vm.count("help-debug")) {
    printUsage(argv[0], allOpts);
    return 0;
  }

  if (vm.count("version")) {
    printVersionInformation();
    return 0;
  }

  if (vm.count("game-root")) {
    gamerootPath = vm["game-root"].as<string>();

    if (!fs::exists(gamerootPath)) {
      cerr << "ERROR: Path '" << gamerootPath << "' does not exist." << endl;
      return -1;
    }

    if (!fs::is_directory(gamerootPath)) {
      cerr << "ERROR: Path '" << gamerootPath << "' is not a directory."
           << endl;
      return -1;
    }

    // Some games hide data in a lower subdirectory.  A little hack to
    // make these behave as expected...
    if (CorrectPathCase(gamerootPath / "Gameexe.ini").empty()) {
      if (!CorrectPathCase(gamerootPath / "KINETICDATA" / "Gameexe.ini")
               .empty()) {
        gamerootPath /= "KINETICDATA/";
      } else if (!CorrectPathCase(gamerootPath / "REALLIVEDATA" / "Gameexe.ini")
                      .empty()) {
        gamerootPath /= "REALLIVEDATA/";
      } else {
        cerr << "WARNING: Path '" << gamerootPath << "' may not contain a "
             << "RealLive game." << endl;
      }
    }
  } else {
    gamerootPath = instance.SelectGameDirectory();
    if (gamerootPath.empty())
      return -1;
  }

  if (vm.count("start-seen"))
    instance.set_seen_start(vm["start-seen"].as<int>());

  if (vm.count("dump-seen"))
    instance.set_dump_seen(vm["dump-seen"].as<int>());

  if (vm.count("memory"))
    instance.set_memory();

  if (vm.count("undefined-opcodes"))
    instance.set_undefined_opcodes();

  if (vm.count("count-undefined"))
    instance.set_count_undefined();

  if (vm.count("trace"))
    instance.set_tracing();

  if (vm.count("load-save"))
    instance.set_load_save(vm["load-save"].as<int>());

  if (vm.count("font"))
    instance.set_custom_font(vm["font"].as<string>());

  instance.Run(gamerootPath);

  return 0;
}
